compiler/rustc_resolve/src/diagnostics.rs

```diff
@@ -1617,7 +1617,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             let post = format!(", consider renaming `{}` into `{snippet}`", suggestion.candidate);
             (span, snippet, post)
         } else {
-            (span, suggestion.candidate.to_string(), String::new())
+            (span, suggestion.candidate.to_ident_string(), String::new())
```

Comment:
This is already an improvement. However, it breaks down when multiple crates are involved whose editions differ. The docs of `to_ident_string` do allude to this footgun (“*symbols don’t keep the rawness flag or edition, so we have to guess the rawness using the global edition.*”).

We can do a bit better since `TypoSuggestion` keeps tracks of the `Span` if the candidate was derived from an `Ident`.

A localized fix would be:

```suggestion
            let sugg = Ident::new(suggestion.candidate, suggestion.span.unwrap_or_default()).to_string();
            (span, sugg, String::new())
```

Reply:
To give you an example were your current fix ceases to work:

```rs
// :::::: a.rs ::::::   rustc a.rs --crate-type=lib --edition=2015
pub fn async() {}

// :::::: b.rs ::::::   rustc b.rs --extern a -L. --edition=2018
fn main() {
    a::asynk(); //~ ERROR cannot find function `asynk` in crate `a`
    //~^ HELP a function with a similar name exists: `async`
}
```

Using the `ui_test`-style `compiletest` format:

```rs
// :::::: auxiliary/a.rs ::::::
pub fn async() {}

// :::::: b.rs ::::::
//@ edition: 2018
//@ aux-crate:a=a.rs

fn main() {
    a::asynk();  //~ ERROR cannot find function `asynk` in crate `a`
    //~^ HELP a function with a similar name exists: `async`
}
```

Please add this as a regression test.

Reply:
However, instead of just fixing this specific instance, could you instead introduce a method on `TypoCandidate` called something like `suggestable_candidate` with a doc comment mentioning that its purpose is to properly escape the candidate with `r#` if necessary and use it not just here but also in other places that currently use `suggestion.candidate` where it makes sense (you can leave all those `suggestion.candidate.as_str()` usage sites as is).

Signature should ideally be `fn suggestable_candidate(&self) -> impl Display` or `-> Ident` if the former doesn't work out.

Reply:
Actually, nvm. I don't know what I wrote there :facepalm: It was early.